pkgs/by-name/vs/vscode-json-languageserver/package.nix

```
@@ -0,0 +1,39 @@
+{
+  lib,
+  buildNpmPackage,
+  fetchFromGitHub,
+  typescript,
+}:
+
+buildNpmPackage rec {
```

```suggestion
buildNpmPackage (finalAttrs: {
```

I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

And then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent examples of implementations: https://github.com/NixOS/nixpkgs/pulls?q=is%3Apr+is%3Aclosed+finalAttrs+pattern+

Feel free to reach out if you need some assistance.

---

fixed.
Thank you for your obvious and details explanation of the `finalAttrs`.
I understand the problem of `rec` and why we should use the `finalAttrs` pattern.